### PR TITLE
Unify scaled INT8 mm

### LINKED ISSUE #862: Unify scaled INT8 matmul
With the new addition of INT8 mixed-precision training, there are now 2 implementations of scaled INT8 matmul (INT8 matmul + dequant)

- https://github.com/pytorch/ao/blob/main/torchao/kernel/intmm_triton.py
- https://github.com/pytorch/ao/blob/main/torchao/prototype/quantized_training/int8_mm.py

I have identified the key differences

`intmm_triton.py` | `int8_mm.py`
---|---
Only fuse act scale | Fuse both act scale and weight scale
Scale step is `acc_i32 x scale` | Scale step is cast to fp32 `acc_i32.to(f32) x scale.to(f32)`
Different autotune configs | Different autotune configs

Ideally we should only keep 1. The tedious part is to validate there is no accuracy+speed regression, regardless of which final implementation we will adopt.

Here are the places that use `intmm_triton.py`

- https://github.com/pytorch/ao/blob/main/benchmarks/intmm.py 
- https://github.com/pytorch/ao/blob/main/torchao/quantization/utils.py (I think this file is legacy from module-swap API?)
- https://github.com/pytorch/ao/blob/main/torchao/dtypes/affine_quantized_tensor.py
- https://github.com/pytorch/ao/blob/main/test/kernel/test_autotuner.py

-> Basically ensure INT8 dynamic quantization for Llama and SAM benchmarks don't regress

Here are the places that use `int8_mm.py`

- https://github.com/pytorch/ao/blob/main/torchao/prototype/quantized_training/int8_mixed_precision.py
- https://github.com/pytorch/ao/blob/main/benchmarks/quantized_training/benchmark_int8mm.py (this is similar to the benchmark script for `intmm_triton.py` above

-> Ensure INT8 mixed-precision training doesn't regress

Another question. Is it ok to change `int_scaled_matmul()` signature to accept scales for both A and B instead of only for A?

### PR DESCRIPTION
**Pull request by gau-nernst**
Fixes #862

### CODE REVIEW

File: torchao/dtypes/affine_quantized_tensor.py [hunk 1]
```diff
@@ -1160,31 +1160,14 @@ def _linear_int8_act_int8_weight_check(input_tensor, weight_tensor, bias):
...
     # 1. do the matrix form of dot(X_i, W_j)
-    #
-    #
     # 2. rescale the output
-    #
-    # in cases with large matrices, y_dot_int32 can grow sufficiently
-    # large that y_dot_int32 * a float16 scale is greater than the maximum
-    # value of a float 16, (which results in a value of inf even if multiplying
-    # by the other scale would bring it within the expected range)
-
```
Comment (gau-nernst) [inline torchao/dtypes/affine_quantized_tensor.py:1173]: overflow issue is now handled inside `int_scaled_matmul()`, which always perform multiplication in FP32.

File: test/kernel/test_autotuner.py [hunk 2]
```diff
@@ -47,17 +46,10 @@ def test_int_mm(self, device, dtype):
...
 
-    @parameterized.expand(
-        [
-            ("cuda", torch.bfloat16),
-            ("cpu", torch.bfloat16),
-            ("cuda", torch.float16),
-            ("cpu", torch.float16),
-        ]
-    )
-    def test_int_scaled_mm(self, device, dtype):
```
Comment (gau-nernst) [inline test/kernel/test_autotuner.py:60]: dtype is always overriden to BF16 later

**Comment by gau-nernst** [comment]
The triton kernel in `torchao.kernel.intmm_triton` is highly coupled to `torchao.kernel.autotuner`, which makes it unnecessary complex to separate them out. `intmm_triton` also has inference in mind, with exhaustive autotuning and cached config, which is not suitable for training. Closing this for now.
